### PR TITLE
ADC: Add support for Microchip MCP3204/MCP3208

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
These commits adds support for the MCP3204/MCP3208 12-bit ADCs from Microchip.

### CODE REVIEW

**Comment by galak** [comment]
DT stuff looks good to me.

File: drivers/adc/adc_mcp320x.c [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+		adc_context_on_sampling_done(&data->ctx, dev);
+	}
+}
+
+static int mcp320x_init(struct device *dev)
+{
+	const struct mcp320x_config *config = dev->config->config_info;
+	struct mcp320x_data *data = dev->driver_data;
+
+	data->dev = dev;
```
Comment (anangl) [inline drivers/adc/adc_mcp320x.c]: Is this actually needed? I couldn't find a place where this field was used.

File: drivers/adc/adc_mcp320x.c [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+			channel_cfg->acquisition_time);
+		return -ENOTSUP;
+	}
+
+	if (channel_cfg->channel_id >= config->channels) {
+		LOG_ERR("unsupported channel id '%d'", channel_cfg->channel_id);
+		return -ENOTSUP;
+	}
+
+	WRITE_BIT(data->differential, BIT(channel_cfg->channel_id),
```
[the remark this answers is not in the document]
Comment (henrikbrixandersen) [reply drivers/adc/adc_mcp320x.c]: Thanks, fixed.

File: drivers/adc/adc_mcp320x.c (hunk 1 above)
Comment (henrikbrixandersen) [reply drivers/adc/adc_mcp320x.c]: No, that was leftover from an earlier implementation. Fixed.
Thank you for reviewing.

**Review by anangl — APPROVED** [review]

File: dts/bindings/iio/adc/microchip,mcp3204.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+description: Microchip MCP3204
```
Comment (erwango) [inline dts/bindings/iio/adc/microchip,mcp3204.yaml]: I might have missed something, but don't we need copyright and SPDX boilerplate in bindings too ?

**Review by erwango — COMMENTED** [review]
Looks good on shield side, but genuine question on yaml binding.

File: dts/bindings/iio/adc/microchip,mcp3204.yaml (hunk 3 above)
Comment (henrikbrixandersen) [reply dts/bindings/iio/adc/microchip,mcp3204.yaml]: YAML files are excluded in the GitHub license check workflow (see 3b625aaaa97b33d224e2c0445c441e481ccc4971 by @nashif).

Comment (erwango) [reply dts/bindings/iio/adc/microchip,mcp3204.yaml]: ... and most of yaml bindings have Licences. Besides, I can't see why there would be a derogation specifically on these files.
So what should be the policy? @MaureenHelm , @nashif

Comment (henrikbrixandersen) [reply dts/bindings/iio/adc/microchip,mcp3204.yaml]: There are approximately 120 files under `dts/bindings/` without license and copyright (and approximately 400 files with license and copyright). I suggest leaving this PR as-is and opening a separate issue for dealing with the policy on licenses for YAML files.

File: boards/shields/mikroe_adc_click/doc/index.rst [hunk 4]
```diff
@@ -0,0 +1,61 @@
+.. _mikroe_adc_click_shield:
+
+MikroElektronika ADC Click
+##########################
+
+Overview
+********
+
+The MikroElektronika ADC Click carries the `MCP3204`_ 12-bit
+Analog-to-Digital converter in a `mikroBUS`_ (tm) form factor.
```
[the remark this answers is not in the document]
Comment (henrikbrixandersen) [reply boards/shields/mikroe_adc_click/doc/index.rst]: @mbolivar-nordic Updated, thanks.

File: dts/bindings/iio/adc/microchip,mcp3204.yaml (hunk 3 above)
Comment (erwango) [reply dts/bindings/iio/adc/microchip,mcp3204.yaml]: Thanks for this counting. Ok to deal with that later.

**Review by erwango — APPROVED** [review]

**Pull request merged**